[privacy] Implement privacy decrypt support in message layer

turon opened the pull request:
#### Issue Being Resolved
* Fixes #19547 for Incoming Messages 
* Superset to #22769

#### Change overview
Adds support for privacy decryption of group messages per spec.
Includes unit test to verify the feature.

mrjerryjohns commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +779,60 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
+
+        if (packetHeader.HasPrivacyFlag())
```
I don't see the changes that implement the outgoing part of message transmission using the privacy flag -if that's not present in the tree right now, any reason why that's not bundled with this PR as well?

turon replied:
I'd like to get 1.0 nodes to do the right thing on the receive side ASAP.  Sending with Pflag = 1 has extra considerations given SVE and 1.0 nodes before this patch won't be able to process such messages.

turon replied:
sending private messages is out-of-scope of this PR.

mrjerryjohns commented on src/transport/tests/TestSessionManagerDispatch.cpp:
```diff
@@ -165,6 +165,35 @@ struct MessageTestEntry theMessageTestVector[] = {
+        .groupId      = 2,
+        .sourceNodeId = 0x0000000000000002ULL,
+    },
+    {
```
What about YAML tests? We should add a functional/end-to-end test to validate the feature here.

turon replied:
This PR just adds decrypt capability for forward compatibility.  All nodes will still generate Pflag=0 messages however, so YAML tests are premature.

mrjerryjohns commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +779,60 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
+            ReturnOnFailure(context.PrivacyDecrypt(privacyHeader, privacyLength, privacyHeader, packetHeader, mac));
+        }
+
+        ReturnOnFailure(packetHeaderCopy.DecodeAndConsume(msgCopy));
```
We need to ensure that older devices running the v1.0 branch don't have unexpected behaviors if a message that has the privacy flag is sent to them.

From what I understand of the existing code, it looks like they call this function today fairly early on in message processing in `SessionManager::OnMessageReceived`, which means they'll have bogus values for the source/dest/message counter fields in the message right?

Ultimately, it seems like that won't be a problem however, since by the time they get to the old logic in this function, it would have failed the call to `SecureMessageCodec::Decrypt`?

All in all, it would be really useful if you could run a test against the 1.0 branch to ensure this actually pans out, and that devices don't just crash.

turon replied:
Yes, nodes running code before this patch will attempt to process garbage header fields.  The best coverage for these scenarios would be fuzzing tests.  I would like this patch to be merged into 1.0 as it is spec mandatory behavior to process privacy headers properly.

turon replied:
Added `CHIP_CONFIG_PRIVACY_ACCEPT_NONSPEC_SVE2` as an option to handle the rare set of erroneous nodes that send with P=1 and no privacy prior to [#22425](https://github.com/project-chip/connectedhomeip/pull/22425/files#diff-76f701cb7a091ad07f50f468176175f53297e3efe37681a4f4205ecce234c93fL169).  Note the cost of enabling that is two rounds of decryption per key both with and without privacy. This PR includes extensive positive and negative unit tests for receiving group messages.

The `packetHeaderCopy` and `msgCopy` fields are designed to handle non-private fields such as message flags and session id and return remaining privacy decrypted fields after authentication.

mrjerryjohns commented on src/transport/raw/MessageHeader.h:
```diff
@@ -316,6 +325,38 @@ class PacketHeader
+            length += sizeof(GroupId);
+        }
+        /*
+        if (mSecFlags.Has(Header::SecFlagValues::kMsgExtensionFlag))
```
Is this commented out for a reason?

turon replied:
Will remove.

mrjerryjohns commented:
This is likely not going to go into the v1.0 branch. How exactly are new clients that want to send privacy-encrypted group messages to older servers running  going to be able to discover which of those servers support this feature vs. which don't?

This seems like it needs a new 'message layer revision' support of sorts.

tcarmelveilleux commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +779,60 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
+    uint8_t * data     = msg->Start();
+    uint16_t len       = msg->DataLength();
+    uint16_t footerLen = packetHeader.MICTagLength();
+    VerifyOrReturn(footerLen <= len); // err = CHIP_ERROR_INVALID_MESSAGE_LENGTH;
```
Why the commented `err = ` ?

turon replied:
Will remove.

turon reviewed (commented):
Thanks for the reviews.

mrjerryjohns commented on src/transport/SessionManager.cpp:
```diff
@@ -736,27 +748,21 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
 
-    GroupId groupId = packetHeader.GetDestinationGroupId().Value();
-
-    if (msg.IsNull())
```
Is this check no longer needed and if so, why?

turon replied:
It moved to line 809 below.

mrjerryjohns replied:
Ah, it got moved down. Nonetheless, why wouldn't we do the check here? Doesn't that ensure that we're not accessing out of bounds memory later below?

turon replied:
No, the check is invalid until `DecodeAndConsume` is called.  Also, the `DecodeFixed` does checks that the minimum sized header exists.  I suppose we could possibly move the check into DecodeFixed and DecodeAndConsume (that create a new assumption that zero payload messages are invalid globally), but the philosophy for this PR was to minimize changes to the existing message processing.

turon replied:
Added IsNull check after msgClone.

tcarmelveilleux reviewed (dismissed):
This has wire format changes not tested in SVE/SVE2, and not covered by test plans.

Adding this late in the game is a risk to future compatibility.

- [ ] Justification for why this is added now, and description of mitigations and integration tests to be done are required before this can be merged even in master.

turon commented:
> This has wire format changes not tested in SVE/SVE2, and not covered by test plans.

The PR includes tests.  SVE and SVE2 had broken behavior with regard to privacy (was setting the Pflag yet not applying privacy).
 
> Adding this late in the game is a risk to future compatibility.

This change actually provides a path to future compatibility whereas the current behavior is not spec compliant.  This patch should be expedited and merged into master ASAP so 1.0 nodes based on the SDK can properly decrypt group messages with privacy enabled.

> * [ ]  Justification for why this is added now, and description of mitigations and integration tests to be done are required before this can be merged even in master.

The PR includes unit tests for properly parsing and decrypting messages both with the privacy flag enabled and disabled.  The PR was posted sometime ago but was blocking on SiLabs platform CI tests which were fixed/merged 2 days ago: #22920.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -656,6 +668,13 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
+    // Drop secure unicast messages with privacy enabled.
+    if (packetHeader.HasPrivacyFlag())
+    {
+        ChipLogError(Inet, "Dropping secure unicast message with privacy flag set");
```
So this PR is only supporting privacy decrypt for group messages?

turon replied:
Yes.  I suppose we could add the processing to unicast as well -- it would only scramble the counters as src / dst are always elided.

bzbarsky-apple replied:
Per spec unicast messages generally have this, no?

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +786,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
+        if (packetHeader.HasPrivacyFlag())
+        {
+            // Perform privacy deobfuscation, if applicable.
+            uint8_t * privacyHeader = msgCopy->Start() + Header::kPrivacyHeaderOffset;
```
Can we make this a (static?) method on PacketHeader or something?  It's weird to be getting the length from the PacketHeader but the data from somewhere else.

Maye DecodeFixed should stash this data and then we can just ask the PacketHeader instance for it?

turon replied:
`DecodeFixed` is actually storing parsed header fields that can be reused for all copies of that message.

The pointer to the start of the privacyHeader is unique to each msgCopy and needs to be dynamically calculated in order to deobfuscate so we can call `DecodeAndConsume`.

bzbarsky-apple replied:
Ah, indeed.  This still feels like it would be better as a static PacketHeader method....

turon replied:
Added a fix.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +786,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
     while (!decrypted && iter->Next(groupContext))
     {
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
```
This is where the `msgCopy.IsNull()` check needs to be, no?  Before we start dereferencing it...

turon replied:
Added extra IsNull check after clone.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +786,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
+
+        ReturnOnFailure(packetHeaderCopy.DecodeAndConsume(msgCopy));
+
+        if (msgCopy.IsNull())
```
This check doesn't make sense here.  If `msgCopy` were null, we would have crashed before this point.

bzbarsky-apple replied:
That does not make sense to me.  `DecodeAndConsume` is not going to make the handle null....

turon replied:
Removed as advised.

bzbarsky-apple commented on src/transport/raw/MessageHeader.cpp:
```diff
@@ -131,6 +131,31 @@ uint16_t PayloadHeader::EncodeSizeBytes() const
+    int version;
+
+    uint8_t msgFlags;
+    SuccessOrExit(err = reader.Read8(&msgFlags).StatusCode());
```
This is duplicating a bunch of code from `Decode`.  Can that move to a helper method called from both, please?  Presumably taking a `LittleEndian::Reader &` as an argument.

bzbarsky-apple commented on src/transport/raw/MessageHeader.h:
```diff
@@ -50,6 +50,13 @@ typedef int PacketHeaderFlags;
 
+enum Consts
+{
+    kHeaderMinLength        = 8,
+    kPrivacyHeaderMinLength = 4,
+    kPrivacyHeaderOffset    = 4,
```
These should be under PacketHeader, no?  They have nothing to do with PayloadHeader....

turon replied:
Moved

turon commented on src/transport/SessionManager.cpp:
```diff
@@ -656,6 +668,13 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
         return;
     }
 
+    // Drop secure unicast messages with privacy enabled.
```
1.0: Sends group P=0.  Decodes P=any as if no privacy.
SVE: Sends group P=1.  Decodes P=any as if no privacy.

The SVE behavior breaks the feature if we want to support compatibility with SVE, the P flag is meaningless and always invalid.

bzbarsky-apple replied:
Yes, I am aware of that.  The question is how we deploy in practice given that pre-1.0 things are shipping...

turon replied:
What I said earlier is incorrect.  SVE took the patch to Send group P=0, and decode P=any as plaintext headers.

turon replied:
This particular check is for secure unicast (PASE / CASE) which do not set P=1 in any circumstance.

woody-apple reviewed (changes requested):
This should be discussed, as we can’t break compatibility with what has shipped

turon commented:
> This should be discussed, as we can’t break compatibility with what has shipped

The set of go forward options in my view are:
1) Defer this feature as provisional.  
    Deprecate to original 0x80 mask P flag. 
    Define a new 0x40 mask Pr flag as part of spec 1.0 errata and post-1.0 provisional work.
    Extend Group Key Management cluster:
       - Add FeatureFlag to signal support of Pr flag by a node.
       - Add flag to Group Security Policy to signal use of Pr flag is enabled for a group.
    Add these changes to SDK and Spec as errata.
    Spec also needs errata to keep Message Extensions (MX) length as plain text and not privacy process extensions. 

2) Toss the Privacy feature altogether.  Remove it from the 1.0 spec as errata.
     Remove extra code paths for calculating privacy keys, etc.

3) Add this privacy parsing support to v1.0 branch as a spec compliance change before v1.0-branch is cut or soon after and encourage device makers that intend to use groupcast to deploy with v1.0-branch.  (This was my original hope and understanding when posting this PR, but the delays and elevation of SVE2 as production have dashed this path as impractical.)

So, given SVE2 is the golden reference, our choice is between path (1) and (2).  Path (1) requires clarity on the process for implementing provisional features. Thoughts?

turon commented:
Added `CHIP_CONFIG_PRIVACY_ACCEPT_NONSPEC_SVE2` option for support of the erroneous pre-SVE2 senders.
Added many unit tests for the various permutations of group messages, including negative test cases.

This adds the ability to privacy decrypt on the receiver side per specification so we can rollout privacy encrypted senders at some future date. @bzbarsky-apple @tennessee-google please take a look.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -568,11 +568,21 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
 }
 
-void SessionManager::UnauthenticatedMessageDispatch(const PacketHeader & packetHeader, const Transport::PeerAddress & peerAddress,
+void SessionManager::UnauthenticatedMessageDispatch(PacketHeader & packetHeader, const Transport::PeerAddress & peerAddress,
```
This "pass a non-const-ref" bit makes it look like the caller cares about the result, but in practice it doesn't, right?

How about:

1. We make this a const ref, but rename to "partialPacketHeader" or some such.
2. The Unicast/Insecure callees then instantiate another PacketHeader called `packetHeader` and decode into that.
3. The group callee is already instantiating another PacketHeader.

Thoughts?  I would be OK with a followup for these changes, because we'd want to see the impact it has on the stack and whatnot.

turon replied:
Reverted back to const and renamed `partialPacketHeader` as suggested.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -656,6 +668,13 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
     }
 
+    // Drop secure unicast messages with privacy enabled.
+    if (packetHeader.HasPrivacyFlag())
```
Shouldn't this happen up front before we DecodeAndConsume, so that does not get garbage data?  Similar to how unsecured messages do it.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -736,27 +755,75 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
+ * Helper function to implement a single attempt to decrypt a groupcast message
+ * using the given group key and privacy setting.
+ *
+ * @param[in] packetHeader The partial packet header of the non-obfuscated fields of the message header
```
I think it's worth documenting what "partial" means here: It's had DecodeFixed called on it.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -773,29 +840,54 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
-        // Optimization to reduce number of decryption attempts
-        if (groupId != groupContext.group_id)
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
```
So we have to make a copy of the message even for non-matching group ids, because we can't apply privacy deobfuscation to get the group id without making the copy?

I wonder whether it's worth, as a followup, seeing whether we can copy just the privacyheader parts.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -736,27 +755,75 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & packetHea
+ * @return true if the message was decrypted successfully
+ * @return false if the message could not be decrypted
+ */
+static bool GroupKeyDecryptAttempt(PacketHeader & packetHeader, PacketHeader & packetHeaderCopy, PayloadHeader & payloadHeader,
```
Which of these references can/should be const?  Pretty sure msgCopy should.  And perhaps mac and groupContext?

bzbarsky-apple commented on src/transport/raw/MessageHeader.h:
```diff
@@ -324,6 +365,30 @@ class PacketHeader
+     *    CHIP_ERROR_INVALID_ARGUMENT on insufficient buffer size
+     *    CHIP_ERROR_VERSION_MISMATCH if header version is not supported.
+     */
+    CHIP_ERROR DecodeFixedCommon(Encoding::LittleEndian::Reader & reader);
```
Shouldn't this be a private API?

robszewczyk commented on src/lib/core/CHIPConfig.h:
```diff
@@ -1169,6 +1169,19 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
+ * This is a temporary workaround for interoperability with those erroneous early-SVE2 implementations.
+ * The cost of this compatibity mode is twice as many decryption steps per received group message.
+ *
+ * TODO: Remove this workaround once interoperability with legacy pre-SVE2 is no longer required.
```
File a tracking issue, perhaps xref in the comment?

turon replied:
Will add ref to #24573
